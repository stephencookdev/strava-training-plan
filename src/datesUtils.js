export const MINUTE_IN_MS = 1000 * 60;
export const DAY_IN_MS = MINUTE_IN_MS * 60 * 24;
export const WEEK_IN_MS = DAY_IN_MS * 7;

export const getStartOfDay = (date) => {
  return Math.floor(date / DAY_IN_MS) * DAY_IN_MS + 1;
};

export const getEndOfDay = (date) => {
  return Math.ceil(date / DAY_IN_MS) * DAY_IN_MS - 1;
};
