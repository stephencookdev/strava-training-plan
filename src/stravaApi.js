export const grabAccessTokens = async () => {
  const queryParams = new URLSearchParams(location.search);
  const authCode = queryParams.get("code");

  if (!authCode) return;

  const authResult = await fetch("/.netlify/functions/auth", {
    method: "POST",
    body: authCode,
  });
  const authResultJson = await authResult.json();

  if (authResultJson.accessToken) {
    window.history.replaceState({}, document.title, "/");

    localStorage.setItem("accessToken", authResultJson.accessToken);
    localStorage.setItem("refreshToken", authResultJson.refreshToken);
  }
};

export const getActivities = async (accessToken, { before, after } = {}) => {
  const queryParams = [];
  if (before) queryParams.push(`before=${Math.floor(before / 1000)}`);
  if (after) queryParams.push(`after=${Math.floor(after / 1000)}`);

  if (after && after > new Date()) {
    // Strava API doesn't allow future `after` dates
    return [];
  }

  const listResponse = await fetch(
    `https://www.strava.com/api/v3/athlete/activities?${queryParams.join("&")}`,
    {
      headers: {
        Authorization: `Bearer ${accessToken}`,
      },
    }
  );
  const list = await listResponse.json();

  return list.map((activity) => {
    return {
      date: new Date(activity.start_date),
      distance: activity.distance,
      movingTime: activity.moving_time,
      isRace: activity.workout_type === 1,
    };
  });
};
