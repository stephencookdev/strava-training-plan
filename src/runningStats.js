import { DAY_IN_MS, WEEK_IN_MS } from "./datesUtils";

const RACE_WEIGHT = 100;
const DAYS_TO_ZERO_WEIGHTING = 7 * 6; // (6 weeks)
const KM_PACE_PER_WEAK_INC = 0.2;
const INITIAL_JUMP = 0.1;

const getTimeWeighting = (activity, targetRace) => {
  const date = activity.date;
  const ZERO_THRESHOLD = DAY_IN_MS * DAYS_TO_ZERO_WEIGHTING;
  const timeOfZero = targetRace.trainingStartDate - ZERO_THRESHOLD;

  const t = Math.max(0, Math.min(1, (date - timeOfZero) / ZERO_THRESHOLD));

  return t * t;
};

const getRaceWeighting = (activity) => {
  return activity.isRace ? RACE_WEIGHT : 1;
};

const getWeeklyMileage = (activities, targetRace) => {
  const sortedActivities = activities.sort((a, b) => a.date - b.date);
  const earliestActivity = sortedActivities[0];
  const latestActivity = sortedActivities[sortedActivities.length - 1];

  const daysBetween = Math.ceil(
    (latestActivity.date - earliestActivity.date) / DAY_IN_MS
  );
  let dayWeightingSum = 0;
  for (let i = 0; i <= daysBetween; i++) {
    dayWeightingSum += getTimeWeighting(
      {
        date: targetRace.trainingStartDate - DAY_IN_MS * i,
      },
      targetRace
    );
  }

  const summedMileage = activities.reduce(
    (acc, cur) => acc + cur.distance * getTimeWeighting(cur, targetRace),
    0
  );

  return (7 * summedMileage) / dayWeightingSum;
};

export const getCurrentPotential = (activities, targetRace) => {
  const cookedActivities = activities.map((activity) => ({
    ...activity,
    weight: getTimeWeighting(activity, targetRace) * getRaceWeighting(activity),
  }));

  const summedActivity = cookedActivities.reduce(
    (acc, activity) => ({
      distance: acc.distance + activity.distance * activity.weight,
      movingTime: acc.movingTime + activity.movingTime * activity.weight,
      weight: acc.weight + activity.weight,
    }),
    { distance: 0, movingTime: 0, weight: 0 }
  );

  const averageActivity = {
    distance: summedActivity.distance / summedActivity.weight,
    movingTime: summedActivity.movingTime / summedActivity.weight,
  };

  const potential = {
    distance: averageActivity.distance,
    movingTime: averageActivity.movingTime,
    pace: averageActivity.movingTime / averageActivity.distance,
    weeklyMileage: getWeeklyMileage(activities, targetRace),
  };

  return potential;
};

export const getPeakReqs = (targetRace) => {
  const peakDistance = 2 * targetRace.distance;

  const targetSecondMeterPace = targetRace.movingTime / targetRace.distance;
  const targetMinuteKmPace = (1000 * targetSecondMeterPace) / 60;
  const peakMinuteKmPace = targetMinuteKmPace + KM_PACE_PER_WEAK_INC;
  const peakSecondMeterPace = (60 * peakMinuteKmPace) / 1000;

  return { distance: peakDistance, pace: peakSecondMeterPace };
};

export const Riegel = {
  R: 1.15, // see e.g. https://www.theguardian.com/lifeandstyle/the-running-blog/2018/feb/15/an-updated-formula-for-marathon-running-success
  getNewTime(newDistance, old) {
    return old.movingTime * Math.pow(newDistance / old.distance, Riegel.R);
  },
  getNewDistance(newTime, old) {
    return old.distance * Math.pow(newTime / old.movingTime, 1 / Riegel.R);
  },
};

export const getWeeklyIncs = (potential, targetPeak, targetRace) => {
  const msUntilRaceFromTrainingStart =
    targetRace.date - targetRace.trainingStartDate;

  const startingMileage =
    (targetPeak.distance - potential.weeklyMileage) * INITIAL_JUMP +
    potential.weeklyMileage;

  const distanceInc = Math.pow(
    targetPeak.distance / startingMileage,
    WEEK_IN_MS / msUntilRaceFromTrainingStart
  );
  const speedInc =
    1 /
    Math.pow(
      targetPeak.pace / potential.pace,
      WEEK_IN_MS / msUntilRaceFromTrainingStart
    );

  return { distanceInc, speedInc };
};
