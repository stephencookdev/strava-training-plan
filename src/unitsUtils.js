import { MINUTE_IN_MS } from "./datesUtils";

export const METERS_IN_KM = 1000;

export const humanDistance = (distance) => {
  const distanceKm = distance / METERS_IN_KM;
  return `${distanceKm.toFixed(1)}km`;
};

export const humanPace = (movingTime, distance) => {
  const secondMeterPace = movingTime / distance;
  const minuteKmPace = (1000 * secondMeterPace) / 60;

  const minuteKmPaceWhole = Math.floor(minuteKmPace);
  const minuteKmPaceSeconds = Math.floor(
    (minuteKmPace - minuteKmPaceWhole) * 60
  )
    .toString()
    .padStart(2, "0");

  const cookedMinuteKmPace = `${minuteKmPaceWhole}:${minuteKmPaceSeconds}`;

  return cookedMinuteKmPace;
};

export const parseHumanPaceKm = (paceKmStr) => {
  const [minutesStr, secondsStr] = paceKmStr.split(":");
  const minutes = parseInt(minutesStr);
  const seconds = parseInt(secondsStr);

  const minuteKmPace = minutes + seconds / 60;
  const secondMeterPace = (60 * minuteKmPace) / METERS_IN_KM;

  return secondMeterPace;
};
