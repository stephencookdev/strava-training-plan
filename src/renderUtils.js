import { humanDistance, humanPace } from "./unitsUtils";
import { getWeeklyIncs } from "./runningStats";

const DAY_MAP = {
  0: "Monday",
  1: "Tuesday",
  2: "Wednesday",
  3: "Thursday",
  4: "Friday",
  5: "Saturday",
  6: "Sunday",
};

export const renderMetaStatsHtml = ({
  potential,
  target,
  riegelRacePrediction,
  targetPeak,
}) => {
  const { distanceInc, speedInc } = getWeeklyIncs(
    potential,
    targetPeak,
    target
  );

  return `
Potential: distance ${humanDistance(potential.distance)} / ${humanPace(
    potential.movingTime,
    potential.distance
  )} pace
Weekly mileage: ${humanDistance(potential.weeklyMileage)}
    
Target pace: ${humanPace(target.movingTime, target.distance)}
Target distance: ${humanDistance(target.distance)}
Riegel pace @ race distance: ${humanPace(
    riegelRacePrediction.timeAtRaceDistance,
    target.distance
  )}
Riegel distance @ race pace: ${humanDistance(
    riegelRacePrediction.distanceAtRacePace
  )}

Distance INC: ${distanceInc.toFixed(2)}
Speed INC: ${speedInc.toFixed(2)}`.trim();
};

const humanWeekActivitiesSoFar = (activitiesOfWeek) => {
  return activitiesOfWeek
    .map((activity) => ` - ${humanActivity(activity)}`)
    .join("\n");
};

const humanPlan = (plan) => {
  return Object.keys(plan)
    .map((dayKey) => `${DAY_MAP[dayKey]}: ${humanActivity(plan[dayKey])}`)
    .join("\n");
};

const humanActivity = (activity) => {
  const parts = [];

  if (activity.runType) parts.push(`(${activity.runType})`);

  if (activity.movingTime && activity.distance)
    parts.push(
      `${humanDistance(activity.distance)} @ ${humanPace(
        activity.movingTime,
        activity.distance
      )}`
    );

  return parts.join(" ");
};

const weekPlanString = (week) => {
  if (week.weekEnd < Date.now()) {
    return (
      "DONE!\n" +
      humanWeekActivitiesSoFar(week.activitiesOfWeek) +
      "\n" +
      `(total ${humanActivity(week.totalActivities)})`
    );
  } else if (week.weekStart > Date.now()) {
    return humanPlan(week.plan);
  } else {
    return (
      humanWeekActivitiesSoFar(week.activitiesOfWeek) +
      "\n" +
      humanPlan(week.plan)
    );
  }
};

export const renderWeekPlanHtml = ({ weeks, region }) => {
  return (
    "Weeks:\n" +
    weeks
      .map((week) => {
        const start = new Date(week.weekStart);
        const dateF = new Intl.DateTimeFormat(region, {
          weekday: "short",
          month: "short",
          day: "numeric",
        });

        const weekSummary = `Week of ${dateF.format(start)} â€” ${humanDistance(
          week.distance
        )} / ${humanPace(week.pace, 1)}`;

        return weekSummary + "\n" + weekPlanString(week);
      })
      .join("\n\n")
  );
};
