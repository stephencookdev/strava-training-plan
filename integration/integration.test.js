import { advanceTo, clear } from "jest-date-mock";
import { getPeakReqs, getCurrentPotential } from "../src/runningStats";
import { generateWeeksPlan } from "../src/weekPlan";
import { parseHumanPaceKm, KM_IN_METERS } from "../src/unitsUtils";
import { DAY_IN_MS } from "../src/datesUtils";
import { renderWeekPlanHtml } from "../src/renderUtils";

const generateRun = (date, distanceKm, paceKmStr, isRace = false) => {
  const distance = distanceKm * KM_IN_METERS;
  const pace = parseHumanPaceKm(paceKmStr);
  const movingTime = distance * pace;

  date = date instanceof Date ? date : new Date(date);
  date.setHours(8);

  return {
    date,
    distance,
    movingTime,
    isRace,
  };
};

describe("Integration Tests", () => {
  beforeEach(() => {
    clear();
  });

  test("New plan, marathon, intermediate", () => {
    advanceTo(new Date("2018-01-01")); // 2018 has Monday, January 1st

    const targetRace = {
      distance: 42195,
      movingTime: 14400000,
      trainingStartDate: new Date("2018-01-01"),
      date: new Date("2018-03-25Z09:00"),
      taper: 12 * DAY_IN_MS,
    };
    const trainingPrefs = {
      0: null, // Monday
      1: "speed", // Tuesday
      2: "recovery", // Wednesday
      3: "recovery", // Thursday
      4: null, // Friday,
      5: null, // Saturday
      6: "long", // Sunday
    };
    const historicalActivities = [
      generateRun("2017-12-25", 5, "6:30"),
      generateRun("2017-12-27", 5, "5:00"),
      generateRun("2017-12-27", 5, "6:30"),
      generateRun("2017-12-31", 12, "5:45"),
    ];
    const sinceTrainingPlanActivities = [];

    const targetPeak = getPeakReqs(targetRace);
    const potential = getCurrentPotential(historicalActivities, targetRace);

    const weeks = generateWeeksPlan(
      targetRace,
      trainingPrefs,
      targetPeak,
      potential,
      sinceTrainingPlanActivities,
      Date.now()
    );

    expect(
      renderWeekPlanHtml({
        weeks,
        region: "en-GB",
      })
    ).toMatchSnapshot();
  });

  test("Half way through accurate week, marathon, intermediate", () => {
    advanceTo(new Date("2018-01-11")); // 2018 has Monday, January 1st

    const targetRace = {
      distance: 42195,
      movingTime: 14400000,
      trainingStartDate: new Date("2018-01-01"),
      date: new Date("2018-03-25Z09:00"),
      taper: 12 * DAY_IN_MS,
    };
    const trainingPrefs = {
      0: null, // Monday
      1: "speed", // Tuesday
      2: "recovery", // Wednesday
      3: "recovery", // Thursday
      4: null, // Friday,
      5: null, // Saturday
      6: "long", // Sunday
    };
    const historicalActivities = [
      generateRun("2017-12-25", 5, "6:30"),
      generateRun("2017-12-27", 5, "5:00"),
      generateRun("2017-12-27", 5, "6:30"),
      generateRun("2017-12-31", 12, "5:45"),
    ];
    const sinceTrainingPlanActivities = [
      generateRun("2018-01-02", 6.7, "6:04"),
      generateRun("2018-01-03", 6.1, "6:04"),
      generateRun("2018-01-04", 6.1, "6:04"),
      generateRun("2018-01-07", 18.3, "5:41"),

      generateRun("2018-01-09", 7.3, "6:04"),
      generateRun("2018-01-10", 6.6, "6:04"),
    ];

    const targetPeak = getPeakReqs(targetRace);
    const potential = getCurrentPotential(historicalActivities, targetRace);

    const weeks = generateWeeksPlan(
      targetRace,
      trainingPrefs,
      targetPeak,
      potential,
      sinceTrainingPlanActivities,
      Date.now()
    );

    expect(
      renderWeekPlanHtml({
        weeks,
        region: "en-GB",
      })
    ).toMatchSnapshot();
  });

  test("Half way through week, long run on different day, marathon, intermediate", () => {
    advanceTo(new Date("2018-01-11")); // 2018 has Monday, January 1st

    const targetRace = {
      distance: 42195,
      movingTime: 14400000,
      trainingStartDate: new Date("2018-01-01"),
      date: new Date("2018-03-25Z09:00"),
      taper: 12 * DAY_IN_MS,
    };
    const trainingPrefs = {
      0: null, // Monday
      1: "speed", // Tuesday
      2: "recovery", // Wednesday
      3: "recovery", // Thursday
      4: null, // Friday,
      5: null, // Saturday
      6: "long", // Sunday
    };
    const historicalActivities = [
      generateRun("2017-12-25", 5, "6:30"),
      generateRun("2017-12-27", 5, "5:00"),
      generateRun("2017-12-27", 5, "6:30"),
      generateRun("2017-12-31", 12, "5:45"),
    ];
    const sinceTrainingPlanActivities = [
      generateRun("2018-01-02", 6.7, "6:04"),
      generateRun("2018-01-03", 6.1, "6:04"),
      generateRun("2018-01-04", 6.1, "6:04"),
      generateRun("2018-01-07", 18.3, "5:41"),

      generateRun("2018-01-09", 7.3, "6:04"),
      generateRun("2018-01-10", 19.8, "5:41"),
    ];

    const targetPeak = getPeakReqs(targetRace);
    const potential = getCurrentPotential(historicalActivities, targetRace);

    const weeks = generateWeeksPlan(
      targetRace,
      trainingPrefs,
      targetPeak,
      potential,
      sinceTrainingPlanActivities,
      Date.now()
    );

    expect(
      renderWeekPlanHtml({
        weeks,
        region: "en-GB",
      })
    ).toMatchSnapshot();
  });
});
