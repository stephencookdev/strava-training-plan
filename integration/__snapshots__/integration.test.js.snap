// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests Half way through accurate week, marathon, intermediate 1`] = `
"Weeks:
Week 1 of Mon, 1 Jan — 37.6km / 5:53
DONE!
 - 6.7km @ 6:03
 - 6.1km @ 6:03
 - 6.1km @ 6:03
 - 18.3km @ 5:41
(total 37.2km @ 5:52)

Week 2 of Mon, 8 Jan — 40.8km / 5:53
 - 7.3km @ 6:03
 - 6.6km @ 6:03
Thursday: (recovery) 6.7km @ 6:07
Sunday: (long) 20.1km @ 5:41

Week 3 of Mon, 15 Jan — 35.3km / 5:53
Tuesday: (speed) 6.4km @ 6:04
Wednesday: (recovery) 5.8km @ 6:04
Thursday: (recovery) 5.8km @ 6:04
Sunday: (long) 17.4km @ 5:41

Week 4 of Mon, 22 Jan — 47.9km / 5:53
Tuesday: (speed) 8.6km @ 6:04
Wednesday: (recovery) 7.8km @ 6:04
Thursday: (recovery) 7.8km @ 6:04
Sunday: (long) 23.5km @ 5:41

Week 5 of Mon, 29 Jan — 51.9km / 5:53
Tuesday: (speed) 9.4km @ 6:04
Wednesday: (recovery) 8.5km @ 6:04
Thursday: (recovery) 8.5km @ 6:04
Sunday: (long) 25.5km @ 5:41

Week 6 of Mon, 5 Feb — 45.0km / 5:53
Tuesday: (speed) 8.1km @ 6:04
Wednesday: (recovery) 7.4km @ 6:04
Thursday: (recovery) 7.4km @ 6:04
Sunday: (long) 22.1km @ 5:41

Week 7 of Mon, 12 Feb — 60.9km / 5:53
Tuesday: (speed) 11.0km @ 6:04
Wednesday: (recovery) 10.0km @ 6:04
Thursday: (recovery) 10.0km @ 6:04
Sunday: (long) 29.9km @ 5:41

Week 8 of Mon, 19 Feb — 66.0km / 5:53
Tuesday: (speed) 12.0km @ 6:04
Wednesday: (recovery) 10.9km @ 6:04
Thursday: (recovery) 10.9km @ 6:04
Sunday: (long) 32.2km @ 5:41

Week 9 of Mon, 26 Feb — 57.2km / 5:53
Tuesday: (speed) 10.3km @ 6:04
Wednesday: (recovery) 9.4km @ 6:04
Thursday: (recovery) 9.4km @ 6:04
Sunday: (long) 28.1km @ 5:41

Week 10 of Mon, 5 Mar — 77.5km / 5:53
Tuesday: (speed) 15.7km @ 6:01
Wednesday: (recovery) 14.8km @ 6:01
Thursday: (recovery) 14.8km @ 6:01
Sunday: (long) 32.2km @ 5:41

Week 11 of Mon, 12 Mar — 40.1km / 5:53
Tuesday: (speed) 7.2km @ 6:04
Wednesday: (recovery) 6.6km @ 6:04
Thursday: (recovery) 6.6km @ 6:04
Sunday: (long) 19.7km @ 5:41

Week 12 of Mon, 19 Mar — 25.5km / 5:53
Tuesday: (speed) 4.6km @ 6:04
Wednesday: (recovery) 4.2km @ 6:04
Thursday: (recovery) 4.2km @ 6:04
Sunday: (long) 12.5km @ 5:41"
`;

exports[`Integration Tests Half way through week, long run on different day, marathon, intermediate 1`] = `
"Weeks:
Week 1 of Mon, 1 Jan — 37.6km / 5:53
DONE!
 - 6.7km @ 6:03
 - 6.1km @ 6:03
 - 6.1km @ 6:03
 - 18.3km @ 5:41
(total 37.2km @ 5:52)

Week 2 of Mon, 8 Jan — 40.8km / 5:53
 - 7.3km @ 6:03
 - 19.8km @ 5:41
Wednesday: (recovery) 6.8km @ 6:05
Thursday: (recovery) 6.8km @ 6:05

Week 3 of Mon, 15 Jan — 35.3km / 5:53
Tuesday: (speed) 6.4km @ 6:04
Wednesday: (recovery) 5.8km @ 6:04
Thursday: (recovery) 5.8km @ 6:04
Sunday: (long) 17.4km @ 5:41

Week 4 of Mon, 22 Jan — 47.9km / 5:53
Tuesday: (speed) 8.6km @ 6:04
Wednesday: (recovery) 7.8km @ 6:04
Thursday: (recovery) 7.8km @ 6:04
Sunday: (long) 23.5km @ 5:41

Week 5 of Mon, 29 Jan — 51.9km / 5:53
Tuesday: (speed) 9.4km @ 6:04
Wednesday: (recovery) 8.5km @ 6:04
Thursday: (recovery) 8.5km @ 6:04
Sunday: (long) 25.5km @ 5:41

Week 6 of Mon, 5 Feb — 45.0km / 5:53
Tuesday: (speed) 8.1km @ 6:04
Wednesday: (recovery) 7.4km @ 6:04
Thursday: (recovery) 7.4km @ 6:04
Sunday: (long) 22.1km @ 5:41

Week 7 of Mon, 12 Feb — 60.9km / 5:53
Tuesday: (speed) 11.0km @ 6:04
Wednesday: (recovery) 10.0km @ 6:04
Thursday: (recovery) 10.0km @ 6:04
Sunday: (long) 29.9km @ 5:41

Week 8 of Mon, 19 Feb — 66.0km / 5:53
Tuesday: (speed) 12.0km @ 6:04
Wednesday: (recovery) 10.9km @ 6:04
Thursday: (recovery) 10.9km @ 6:04
Sunday: (long) 32.2km @ 5:41

Week 9 of Mon, 26 Feb — 57.2km / 5:53
Tuesday: (speed) 10.3km @ 6:04
Wednesday: (recovery) 9.4km @ 6:04
Thursday: (recovery) 9.4km @ 6:04
Sunday: (long) 28.1km @ 5:41

Week 10 of Mon, 5 Mar — 77.5km / 5:53
Tuesday: (speed) 15.7km @ 6:01
Wednesday: (recovery) 14.8km @ 6:01
Thursday: (recovery) 14.8km @ 6:01
Sunday: (long) 32.2km @ 5:41

Week 11 of Mon, 12 Mar — 40.1km / 5:53
Tuesday: (speed) 7.2km @ 6:04
Wednesday: (recovery) 6.6km @ 6:04
Thursday: (recovery) 6.6km @ 6:04
Sunday: (long) 19.7km @ 5:41

Week 12 of Mon, 19 Mar — 25.5km / 5:53
Tuesday: (speed) 4.6km @ 6:04
Wednesday: (recovery) 4.2km @ 6:04
Thursday: (recovery) 4.2km @ 6:04
Sunday: (long) 12.5km @ 5:41"
`;

exports[`Integration Tests New plan, marathon, intermediate 1`] = `
"Weeks:
Week 1 of Mon, 1 Jan — 37.6km / 5:53
Tuesday: (speed) 6.8km @ 6:04
Wednesday: (recovery) 6.2km @ 6:04
Thursday: (recovery) 6.2km @ 6:04
Sunday: (long) 18.5km @ 5:41

Week 2 of Mon, 8 Jan — 40.8km / 5:53
Tuesday: (speed) 7.4km @ 6:04
Wednesday: (recovery) 6.7km @ 6:04
Thursday: (recovery) 6.7km @ 6:04
Sunday: (long) 20.0km @ 5:41

Week 3 of Mon, 15 Jan — 35.3km / 5:53
Tuesday: (speed) 6.4km @ 6:04
Wednesday: (recovery) 5.8km @ 6:04
Thursday: (recovery) 5.8km @ 6:04
Sunday: (long) 17.4km @ 5:41

Week 4 of Mon, 22 Jan — 47.9km / 5:53
Tuesday: (speed) 8.6km @ 6:04
Wednesday: (recovery) 7.8km @ 6:04
Thursday: (recovery) 7.8km @ 6:04
Sunday: (long) 23.5km @ 5:41

Week 5 of Mon, 29 Jan — 51.9km / 5:53
Tuesday: (speed) 9.4km @ 6:04
Wednesday: (recovery) 8.5km @ 6:04
Thursday: (recovery) 8.5km @ 6:04
Sunday: (long) 25.5km @ 5:41

Week 6 of Mon, 5 Feb — 45.0km / 5:53
Tuesday: (speed) 8.1km @ 6:04
Wednesday: (recovery) 7.4km @ 6:04
Thursday: (recovery) 7.4km @ 6:04
Sunday: (long) 22.1km @ 5:41

Week 7 of Mon, 12 Feb — 60.9km / 5:53
Tuesday: (speed) 11.0km @ 6:04
Wednesday: (recovery) 10.0km @ 6:04
Thursday: (recovery) 10.0km @ 6:04
Sunday: (long) 29.9km @ 5:41

Week 8 of Mon, 19 Feb — 66.0km / 5:53
Tuesday: (speed) 12.0km @ 6:04
Wednesday: (recovery) 10.9km @ 6:04
Thursday: (recovery) 10.9km @ 6:04
Sunday: (long) 32.2km @ 5:41

Week 9 of Mon, 26 Feb — 57.2km / 5:53
Tuesday: (speed) 10.3km @ 6:04
Wednesday: (recovery) 9.4km @ 6:04
Thursday: (recovery) 9.4km @ 6:04
Sunday: (long) 28.1km @ 5:41

Week 10 of Mon, 5 Mar — 77.5km / 5:53
Tuesday: (speed) 15.7km @ 6:01
Wednesday: (recovery) 14.8km @ 6:01
Thursday: (recovery) 14.8km @ 6:01
Sunday: (long) 32.2km @ 5:41

Week 11 of Mon, 12 Mar — 40.1km / 5:53
Tuesday: (speed) 7.2km @ 6:04
Wednesday: (recovery) 6.6km @ 6:04
Thursday: (recovery) 6.6km @ 6:04
Sunday: (long) 19.7km @ 5:41

Week 12 of Mon, 19 Mar — 25.5km / 5:53
Tuesday: (speed) 4.6km @ 6:04
Wednesday: (recovery) 4.2km @ 6:04
Thursday: (recovery) 4.2km @ 6:04
Sunday: (long) 12.5km @ 5:41"
`;
